pkg/operator/k8sutil/network.go

```diff
@@ -32,35 +35,155 @@ const (
+	}
+	// used by 'whereabouts' (legacy)
+	Range string `json:"range,omitempty"`
+
+	// RangeStart string `json:"rangeStart,omitempty"`
+	// RangeEnd   string `json:"rangeEnd,omitempty"`
+	// Routes []struct {
+	// 	Dst string `json:"dst,omitempty"`
+	// } `json:"routes,omitempty"`
+	// Gateway string `json:"gateway,omitempty"`
+}
+
+func (i *IpamConfig) IsEmpty() bool {
+	return reflect.DeepEqual(*i, IpamConfig{})
+}
+
+func GetNetworkAttachmentRanges(nad NetworkAttachmentConfig) ([]string, error) {
+	// TODO: check CNI version; must be 0.2.0+ and less than 2.0.0
+
+	if len(nad.Plugins) == 0 {
+		if nad.Type == "" || nad.Ipam.IsEmpty() {
+			// error
+		}
+
+		ranges, err := GetIpamRanges(nad.Ipam)
+		if err != nil {
+			return []string{}, errors.Wrapf(err, "failed to get address range(s) from NetworkAttachmentDefinition's IPAM config")
+		}
+		return ranges, nil
+	}
+
+	if nad.Type != "" || !nad.Ipam.IsEmpty() {
+
+	}
+
+}
+
+func GetIpamRanges(ipam IpamConfig) ([]string, error) {
+	cidrs := []string{}
+
+	switch ipam.Type {
+	case hostLocalIPAMType:
+		if ipam.Subnet != "" {
+			// because IPAM plugins have been known to change their config formats, and because
+			// config fields could theoretically change meaning, it is good to check that the value
+			// we are grabbing is actually a CIDR and not something different
+			if _, _, err := net.ParseCIDR(ipam.Subnet); err != nil {
+				return cidrs, errors.Wrapf(err, "%q IPAM subnet is invalid", hostLocalIPAMType)
+			}
+			cidrs = append(cidrs, ipam.Subnet)
+		}
+		for _, netRanges := range ipam.Ranges {
+			for _, netRange := range netRanges {
+				if _, _, err := net.ParseCIDR(netRange.Subnet); err != nil {
+					// unique error message for range from new format
+					return cidrs, errors.Wrapf(err, "%q IPAM range subnet is invalid", hostLocalIPAMType)
+				}
+				cidrs = append(cidrs, netRange.Subnet)
+			}
+		}
+		return cidrs, nil
+
+	case staticIPAMType:
+		for _, subnet := range ipam.Addresses {
+			if _, _, err := net.ParseCIDR(subnet.Address); err != nil {
+				return cidrs, errors.Wrapf(err, "%q IPAM subnet address (range) is invalid", staticIPAMType)
+			}
+			cidrs = append(cidrs, subnet.Address)
+		}
+		return cidrs, nil
+
+	case whereaboutsIPAMType:
+		if ipam.Range != "" {
+			if _, _, err := net.ParseCIDR(ipam.Range); err != nil {
+				return cidrs, errors.Wrapf(err, "%q IPAM range is invalid", whereaboutsIPAMType)
+			}
+			cidrs = append(cidrs, ipam.Range)
+		}
+		for _, ipRange := range ipam.IpRanges {
+			if _, _, err := net.ParseCIDR(ipRange.Range); err != nil {
+				// unique error message for range from new format
+				return cidrs, errors.Wrapf(err, "%q IPAM IP range is invalid", whereaboutsIPAMType)
+			}
+			cidrs = append(cidrs, ipRange.Range)
+		}
+		return cidrs, nil
+
+	default:
+		return cidrs, errors.Errorf("cannot automatically read network CIDRs (IP address ranges) from IPAM plugin %q, please specify the network ranges manually", ipam.Type)
+	}
```

Comment:
I'm starting to consider that this is probably the wrong approach. I think we might be better of using a canary architecture like we do with the monitors to apply each NAD to a canary pod and then read the CIDRs out of the pod annotation instead. The various CNI specs linked [here](https://www.cni.dev/docs/spec/#released-versions) break down the requirements for CNI plugins and a standardized method of reporting the address ranges. I fear that continuing to go down the road of analyzing the NAD itself is merely descent into madness.

Reply:
After chatting with Multus/Whereabouts dev, I think the canary process is going to work for us. There are some pretty unlikely scenarios where CIDRs could change from node-to-node. Specifically the host-local CNI setup tends to set up different CIDRs on a per-node basis. However, because Rook expects a single NetworkAttachmentDefinition for the whole Ceph cluster, across all nodes, it makes it much less likely that someone will be able to configure a NAD with a CNI chain that give different CIDRs per-node. 

Given this known caveat, I wonder if we should consider forcing users who use host-local NADs to provide CIDRs in the CephCluster.Network spec manually. @travisn ?

We should definitely document that users with NADs that have per-node or per-zone behavior should specify the CIDRs manually in the CephCluster.Network spec.

The good thing about the canary procedure is that we allow Multus and all the CNI plugins to do their thing, and **then** get the CIDR from there, rather than trying to guess at it based on the NAD. And, the `*/network-status` annotation that Rook would use to get the CIDR(s) is strongly-typed in the CNI spec. There isn't risk that different CNIs would use different fields to get the address.

I'm going to pivot to the canary design and try to get some initial testing in, and I think I have the DNS address of a DHCP server that I can use to test (for personal use only) how DHCP reports CIDRs.